[yell.ow] 지하철 서비스 step3 리뷰 부탁드립니다!

isangu opened the pull request:
안녕하세요! 이번 지하철 서비스 미션을 제출한 옐입니다.
이번 미션은 인증 부분을 구현하는 데에 어려움이 있었습니다.. 😅

- 초기 PathController의 로그인/비로그인 여부를 Authorization Header를 포함하는 경우와 아닌 경우로 구분하여 구현하려고 했습니다. 
실제 구글 크롬에서 테스트 하는 과정에서 비로그인 시에도 Authorization : Bearer null 과 같은 헤더 정보가 포함되어 요청이 와서 AuthenticationPrincipal(required = false)로 설정하여 loginMember의 null 체크 후 로그인/비로그인 여부를 구분하도록 구현하게 되었습니다. 괜찮은 구현 방식일까요?

부족한 부분에 대한 많은 피드백 부탁드립니다! 감사합니다 😆

kakao-yell-ow committed:
feat : 1단계 경로 조회 기능 구현

- PathService, Graph, Path 생성
- Lines 일급 컬렉션 생성
- Section의 간선 역할을 위해 DefaultWeightedEdge 상속
- PathController 최단 경로를 리턴하는 findPath 메서드 추가

kakao-yell-ow committed:
feat : 2단계 회원과 인증 관련 기능 구현

- README.md 파일에 요구사항 추가
- AuthService 인증 기능 구현
- AuthController login 메서드 구현
- AuthenticationPrincipalArgumentResolver resolveArgument 구현
- MemberService findByEmail 메서드 추가
- MemberDao CRUD 예외 상황 처리 로직 구현
- MemberController Error Handling 기능 추가

kakao-pavlo-v committed:
feat : 즐겨찾기 기능 구현

- FavoriteService, FavoriteDao, Favorite 추가
- FavoriteResponse 생성자 추가
- FavoriteController 에서 CRD API 구현
- 즐겨찾기 생성됨 검증 테스트 코드 추가

kakao-pavlo-v committed:
feat : 3단계 거리에 따른 요금 계산 기능 구현

- Section에 LineId 멤버변수 추가
- Section 변경에 따른 LineService, LineDao, SectionDao, Section, Sections 수정
- LineResponse getExtraFare 메서드 추가
- LoginMember 연령에 따른 할인율 계산 기능 추가
- PathService 로그인 비로그인에 따른 요금 계산 기능 구현
- Graph 구간에 포함된 노선 조회 기능 추가
- Path 요금 계산 기능 구현
- PathResponse 정적 팩토리 메소드 오버로딩
- PathController 로그인에 해당하는 경로조회 API 구현
- FareUtil 거리에 해당하는 요금 계산하는 클래스 구현

- DataLoader 노선 생성자 형식 변경
- 테스트 코드 수정

kakao-yell-ow committed:
refactor : 메시지 및 정수 상수화 진행

kakao-yell-ow committed:
refactor : CustomException 생성 및 GlobalControllerAdvice 로 Error Handling 하도록 수정

kakao-yell-ow committed:
refactor : PathController GetMapping 로직변경 및 PathService null 처리 로직 추가

kakao-yell-ow committed:
refactor : 주석 및 TODO 제거

kakao-yell-ow committed:
refactor : 메서드명 변경 및 LoginMember 정보 상수화

kakao-yell-ow committed:
refactor : FareUtil BaseMoney이동 및 LoginMember 나이 계산 부등호 수정

dongjoon1251 commented on README.md:
```diff
@@ -52,3 +52,114 @@ npm run dev
 ## 📝 License
 
 This project is [MIT](https://github.com/next-step/spring-subway-admin-kakao/blob/master/LICENSE) licensed.
+
```
기능과 커밋로그를 잘 남겨주셨네요 💯

dongjoon1251 commented on src/main/java/subway/line/application/LineService.java:
```diff
@@ -36,17 +38,15 @@ private Section addInitSection(Line line, LineRequest request) {
+            Section section = new Section(upStation, downStation, line.getId(), request.getDistance());
+            return sectionDao.insert(section);
         }
         return null;
```
null 을 리턴해서 전달하기 보다는 Optional.ofNullable을 사용해보는게 어떨까요?

dongjoon1251 commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -20,10 +24,17 @@ public boolean supportsParameter(MethodParameter parameter) {
+        HttpServletRequest httpServletRequest = (HttpServletRequest) webRequest.getNativeRequest();
+        String accessToken = AuthorizationExtractor.extract(httpServletRequest);
+        if(accessToken == null || accessToken.equals("null")) {
+            return null;
```
이부분을 질문하신것으로 파악됩니다. 
null 체크는 최대한 지양하는 것이 좋습니다.
항상 LoginMember를 리턴하고 로그인 여부를 필드로 가지거나 로그인되지 않은 사용자를 표현하는 다른 클래스로 표현해보면 어떨까요? 고민해보시고 다른 더 좋은 방법이 떠오르신다면 적용해주세요 :)

dongjoon1251 commented on src/main/java/subway/exceptions/AuthException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class AuthException extends IllegalArgumentException{
+    public AuthException(String message) {
+        super(message);
```
커스텀 예외클래스 👍 
해당 예외가 가지는 message가 항상 같다면 message를 내부의 상수로 두고 사용하면 어떨까요?

dongjoon1251 commented on src/main/java/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,47 @@
+        this.stationService = stationService;
+    }
+
+    public FavoriteResponse createFavorite(LoginMember loginMember, FavoriteRequest favoriteRequest) {
```
request를 바로 넘겨주기보다는 파라미터가 좀 늘어나더라도 필요한 필드만 service layer로 전달하도록 리팩토링해보면 어떨까요?

dongjoon1251 commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public Optional<Favorite> findById(Long id) {
+        List<Favorite> result = jdbcTemplate.query("select id, member_id, source_station_id, target_station_id from FAVORITE where id = ?", rowMapper, id);
```
쿼리 문자열을 상수로 분리해보면 어떨까요?

dongjoon1251 commented on src/main/java/subway/line/domain/Lines.java:
```diff
@@ -0,0 +1,38 @@
+
+    public List<Station> getUniqueStations() {
+        Set<Station> uniqueStation = new HashSet<>();
+        for (Line line : lines) {
```
addAll대신 stream의 flatMap을 사용해보면 어떨까요?

dongjoon1251 commented on src/main/java/subway/line/domain/Lines.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public int getTotalExtraFares(){
+        return lines.stream().mapToInt(Line::getFare).sum();
```
stream을 사용할 때는 가독성을 위해서 한라인에 한포인트만 남도록 줄바꿈해주세요 :)

dongjoon1251 commented on src/main/java/subway/line/domain/Section.java:
```diff
@@ -39,4 +42,39 @@ public Station getDownStation() {
+        return Objects.hash(id, upStation, downStation, distance);
+    }
+
+    @Override
```
현업에서는 Object를 리턴하기보다는 generic을 사용해서 type을 최대한 명시해줍니다.
미션에 기본으로 사용된 라이브러리라서 수정하실 필요는 없고, 참고차 코멘트 남깁니다 :)

dongjoon1251 commented on src/main/java/subway/member/ui/MemberController.java:
```diff
@@ -1,60 +1,63 @@
 import java.net.URI;
 
 @RestController
+@RequestMapping("/members")
```
공통 패스 분리 💯

dongjoon1251 commented on src/main/java/subway/path/domain/Graph.java:
```diff
@@ -0,0 +1,51 @@
+            return new ArrayList<>(dijkstraShortestPath.getPath(sourceStation, targetStation)
+                    .getEdgeList().stream()
+                    .map(Section::getLineId)
+                    .collect(Collectors.toSet()));
```
중복을 제거하기 위해 toSet을 쓰신거라면 distinct 와 같은 메소드를 적용하고 toList로 collect해보면 어떨까요?

dongjoon1251 reviewed (changes requested):
패키지도 잘 나누어 주시고, 각 클래스의 역할도 잘 분리해주셔서 코멘트 드릴 부분이 많지 않네요 👍 
다만 도메인과 관련된 테스트 코드를 추가해주시면 좋을 것 같습니다. 
질문 주신 부분은 코드상에 코멘트로 남겼습니다.

kakao-yell-ow committed:
refactor : AuthException, LineDao, SectionDao, Lines 리팩터링

- AuthException의 메시지 상수화 적용, 상속받은 InvalidToken, UnauthorizedUser, NotExistsDate Exception 오류 메시지 통일
- LineDao, SectionDao 에 Optional 적용, 서비스 레이어의 코드 변경
- Lines의 Stream 가독성 개선

kakao-yell-ow committed:
refactor : FavoriteDao, Lines, Graph 리팩터링

- FavoriteDao Query 상수화
- Lines Stream 로직 flatMap을 사용하여 구현
- Graph Stream 로직 toSet이 아닌 distinct toList를 사용하여 구현

kakao-yell-ow committed:
refactor : FavoriteService, FavoriteController, Graph, Path, PathResponse 리팩터링

- FavoriteService의 createFavorite이 Request가 아닌 필요한 필드를 가지도록 수정
- FavoriteController가 FavoriteRequest가 아닌 Source, Target을 넘겨주도록 수정
- 도메인 객체 Path가 List<StationResponse>가 아닌 List<Station>을 가지도록 수정, 변경에 따른 Graph수정 및 PathResponse 팩토리 메서드 리팩터링

kakao-yell-ow committed:
refactor : 사용하지 않는 패키지 삭제 및 GraphTest, PathTest 도메인 테스트 추가

kakao-yell-ow committed:
refactor : PathController 및 AuthenticationPrincipalArgumentResolver 리팩터링

- LoginMember에 loggedIn 필드 추가 및 경로 비로그인 조회 시 loggedIn=false 로 리턴
- AuthenticationPrincipalArgumentResolver 항상 LoginMember를 리턴하도록 수정
- PathController의 LoginMember loggedIn 값에 따라 요청 분기 하도록 수정

idealgu commented:
피드백 감사드립니다!! 😆
말씀해 주셨던 부분 반영해서 리팩토링을 진행해보았습니다.
처음 질문 드렸던 null 체크는 좋은 방법이 떠오르지 않아 LoginMember에 loggedIn 필드를 추가해 이를 확인하도록 적용해보았습니다.. 😢

dongjoon1251 reviewed (approved):
피드백을 잘 반영해주셨네요:)
Optional도 많은 부분에 잘 적용해주셨습니다.
추가로 코멘트를 드리자면, 테스트 코드에 실패하는 케이스들을 더 다양하게 적용해보면 좋을 것 같습니다.
이번 미션은 여기서 마무리하도록 하겠습니다. 
고생하셨습니다.
